Add code samples to provider page

morganmspencer opened the pull request:
# Overview

This PR adds a tab for code samples on the individual provider's page.

Closes #42 

- Adds toggle button
- Uses markdown files for the code
- Automatically pulls `.md` file if one exists
- Adds [highlight.js](https://highlightjs.org/) for styling
- Uses same highlight style as other Redwood sites
- Help wanted message for providers that do not have code

## To Do

- [x] Add code for each provider
- [x] Update CONTRIBUTING.md

## Screenshots

![image](https://user-images.githubusercontent.com/10109983/115288303-4b0af700-a11f-11eb-932b-e2d0778abcf6.png)

![image](https://user-images.githubusercontent.com/10109983/115288392-60802100-a11f-11eb-9538-6bbd0ba1a584.png)

Tobbe commented on web/src/components/CodeSample/CodeSample.js:
```diff
@@ -0,0 +1,59 @@
+import hljs from 'highlight.js'
+import hljsDefineGraphQL from 'highlightjs-graphql'
+hljsDefineGraphQL(hljs)
+const md = require('markdown-it')({
```
I'm a big fan of remark, any reason you used markdown-it?

morganmspencer replied:
Markdown-it is used on redwoodjs.com so I actually just copied all the settings for it. Not opposed to switching to remark, just tried to maintain consistency

Tobbe replied:
Thought so. Consistency is good. Keep it 👍 
Unless @dthyresson has a strong opinion

dthyresson replied:
Nope! Let's keep for now unless there's a feature in remark we need.

Tobbe commented on web/src/components/CodeSample/CodeSample.js:
```diff
@@ -0,0 +1,59 @@
+      {markdownCode ? (
+        <div
+          className="markdown mt-6"
+          dangerouslySetInnerHTML={{ __html: markdownCode }}
```
With remark you don't have to use `dangerouslySetInnerHTML`

dthyresson replied:
Resolving since decided not to use remark for moment.

Tobbe commented on web/src/components/CodeSample/CodeSample.js:
```diff
@@ -0,0 +1,59 @@
+      }
+    }
+    getMarkdown()
+  })
```
```suggestion
  }, [provider.slug])
```

morganmspencer commented:
@Tobbe and @dthyresson, for the code samples, should I include everything (AuthProvider, config, etc.) or just the utilization of the functions (logIn, logOut, etc.)?

It seems like having the first might not be necessary because that is all generated by RW.

morganmspencer commented:
This is technically ready for review, I just couldn't quite wrap my head around adding the copy button to each potential `pre` block. That could be broken out into a separate issue.

Definitely needs checks on the code samples to make sure they are correct.

Tobbe commented on web/src/lib/code-samples/auth0.md:
```diff
@@ -0,0 +1,30 @@
+        {isAuthenticated ? 'Log Out' : 'Log In'}
+      </button>
+      {!isAuthenticated && (
+        <button
+          onClick={() => {
+            signUp()
+          }}
+        >
```
```suggestion
        <button onClick={signUp}>
```

Just a code style preference from my side here. Let's see what @dthyresson says before changing anything

Tobbe commented on web/src/lib/code-samples/firebase.md:
```diff
@@ -0,0 +1,98 @@
+      ) : (
+        <>
+          {isAuthenticated ? (
+            <button className="btn" onClick={() => logOut()}>
```
```suggestion
            <button className="btn" onClick={logOut}>
```

Tobbe commented:
> It seems like having the first might not be necessary because that is all generated by RW.

I agree. What you have now is good.

dthyresson commented:
> I just couldn't quite wrap my head around adding the copy button to each potential `pre` block. That could be broken out into a separate issue.

@morganmspencer That sounds good. If you want to add that issue and tackle separately (or have otehrs work on it) that makes sense.

I accepted the deploy preview here: https://607e2f16e3f5160007178d99--redwood-playground-auth.netlify.app/

Only thing I noticed was that the Magic Link logo was missing:

![image](https://user-images.githubusercontent.com/1051633/115413859-8bbc4c00-a1c3-11eb-8a25-0b9523edbbb3.png)

morganmspencer commented:
> Only thing I noticed was that the Magic Link logo was missing

My fault, I updated the provider images to just pull from the images folder if it exists. That way we don't need to explicitly define the provider image in `providers.js`, similar to the code samples. I think MagicLink and Azure were camelcased file names so it skipped them.

I've fixed this by adding a lowercase requirement for the image files in the latest commit.

morganmspencer commented:
@dthyresson do you think I should add the RW setup commands above each code example as well?

dthyresson commented:
> do you think I should add the RW setup commands above each code example as well?

I was wondering about that and also the client init (with the AuthProvider etc).

But perhaps it is best to just link to the documentation per provider:

* https://redwoodjs.com/docs/authentication#auth0
* https://redwoodjs.com/docs/authentication#supabase

etc

Perhaps the provider data has a "docsUrl" or something like that to store the link ... and some supporting wording: "How to setup ..."?

morganmspencer commented:
Added the docsUrl for each provider and implemented it for the code samples

dthyresson commented:
Deploy preview is up. LGTM. I'm merging.
